File: services/graphql-billing/src/models/surcharges.rs
use std::sync::Arc;

use async_graphql::dataloader::Loader;
use chrono::{DateTime, NaiveDate, Utc};
use graphql_core::PostgresDataLoader;
use rust_decimal::Decimal;
use uuid::Uuid;

use super::sea_orm_active_enums::SurchargeCalculationMethodEnum;

#[derive(Debug, Clone, Copy, PartialEq, Hash, Eq)]
pub struct PrimaryKey(pub Uuid);

#[derive(Clone, Debug, PartialEq, Eq, async_graphql::SimpleObject, sqlx::FromRow)]
#[graphql(name = "BillingSurcharges")]
pub struct Model {
    pub id: Uuid,
    pub name: String,
    pub r#type: String,
    pub amount: Decimal,
    pub calculation_method: SurchargeCalculationMethodEnum,
    pub is_active: Option<bool>,
    pub valid_from: Option<NaiveDate>,
    pub valid_to: Option<NaiveDate>,
    pub description: Option<String>,
    pub created_at: Option<DateTime<Utc>>,
    pub updated_at: Option<DateTime<Utc>>,
}

impl Loader<PrimaryKey> for PostgresDataLoader {
    type Error = Arc<sqlx::Error>;
    type Value = Model;

    async fn load(
        &self,
        keys: &[PrimaryKey],
    ) -> Result<std::collections::HashMap<PrimaryKey, Self::Value>, Self::Error> {
        let keys = keys.iter().map(|k| k.0).collect::<Vec<_>>();

        let results =
            sqlx::query_as::<_, Self::Value>("select * from billing.surcharges where id = ANY($1)")
                .bind(&keys)
                .fetch_all(&self.pool)
                .await?
                .into_iter()
                .map(|model| (PrimaryKey(model.id), model))
                .collect::<_>();

        Ok(results)
    }
}

File: src/graphql/queries/billing/surcharges.ts
import { graphql } from "@/lib/graphql/client";

// ============================================================================
// SURCHARGE OPERATIONS
// ============================================================================

/**
 * Fetches a single surcharge by its ID.
 * @param id The UUID of the surcharge.
 */
export const getSurcharge = graphql(`
  query GetSurcharge($id: UUID!) {
    billing {
      surcharge(id: $id) {
        id
        name
        type
        amount
        calculationMethod
        isActive
        validFrom
        validTo
        description
        createdAt
        updatedAt
      }
    }
  }
`);

/**
 * Fetches a paginated list of surcharges.
 * @param limit The maximum number of items to return.
 * @param page The page number to retrieve.
 */
export const getSurcharges = graphql(`
  query GetSurcharges($limit: Int!, $page: Int!) {
    billing {
      surcharges(limit: $limit, page: $page) {
        id
        name
        type
        amount
        calculationMethod
        isActive
        validFrom
        validTo
        description
        createdAt
        updatedAt
      }
    }
  }
`);
